local-cluster/tests/local_cluster.rs

```diff
@@ -5012,3 +5027,401 @@ fn test_boot_from_local_state() {
+        // where we disallow resetting to a slot which matches the last vote slot, we still don't build off `D`,
+        // and continue building on `A`.
+        //
+        // The `target_switch_fork_validator_pubkey` fork is necessary in 2. to force the validator stall trying to switch
+        // vote on that other fork and prevent the validator from making a freebie vote from `A` and allowing consensus to continue.
+
+        // It's important we give the `duplicate_fork_validator1_pubkey` very few leader slots so that:
+        // 1. We have ample time to ensure he doesn't have a chance to make a block until after 2 when they see the block is duplicate.
+        // Otherwise, they'll build the block on top of the duplicate block, which will possibly include a vote for the duplicate block.
+        // We want to avoid this because this will make fork choice pick the duplicate block.
+        // 2. Ensure the `duplicate_fork_validator1_pubkey` sees the target switch fork before it can make another vote
+        // on any forks he himself generates from A. Otherwise, he will make a freebie vote on his own fork from `A` and
+        // consensus will continue on that fork.
+
+        // Give the duplicate fork validator plenty of leader slots after the initial delay to prevent
+        // 1. Switch fork from getting locked out for too long
+        // 2. A lot of consecutive slots in which to build up lockout in tower and make new roots
+        // to resolve the partition
+        (duplicate_fork_validator1_pubkey, 500),
+    ];
+
+    let leader_schedule = create_custom_leader_schedule(validator_to_slots.into_iter());
+
+    // 1) Set up the cluster
+    let (duplicate_slot_sender, duplicate_slot_receiver) = unbounded();
+    let (mut cluster, _validator_keypairs) = test_faulty_node(
+        BroadcastStageType::BroadcastDuplicates(BroadcastDuplicatesConfig {
+            partition: ClusterPartition::Pubkey(vec![
+                duplicate_fork_validator1_pubkey,
+                duplicate_fork_validator2_pubkey,
+            ]),
+            duplicate_slot_sender: Some(duplicate_slot_sender),
+        }),
+        node_stakes,
+        Some(validator_keypairs),
+        Some(FixedSchedule {
+            leader_schedule: Arc::new(leader_schedule),
+        }),
+    );
+
+    // 2) Wait for a duplicate slot to land on both validators and for the target switch
+    // fork validator to get another version of the slot. Also ensure all versions of
+    // the block are playable
+    let mut dup_slot;
+    loop {
+        dup_slot = duplicate_slot_receiver
+            .recv_timeout(Duration::from_millis(30_000))
+            .expect("Duplicate leader failed to make a duplicate slot in allotted time");
+        // Make sure both validators received and replay the complete blocks
+        let dup_frozen_hash1 = wait_for_duplicate_fork_frozen(
```

Comment:
`wait_for_duplicate_fork_frozen` might spin forever right? it's expected that some of these blocks could be dead?

Reply:
So the leader is making a duplicate block every `N` blocks where only the last shred is different, but regardless of which version of the shred you get, both versions are playable.

So when the first duplicate is generated there are two cases:

1. My dup validator receives the same version as the leader because one of the other validators sent the original version over turbine faster than the dup leader could send the dup shred. In this case the `original_frozen_hash != dup_frozen_hash` fails and the loop continues until we get the next dup. All slots in between will be playable because we have the original version, and the duplicate leader is making those new blocks based off the original version, not the duplicate version
2. We get a different version than the original leader, in which case we successfully exit the loop

Reply:
got it, both versions are playable so you should never fail to freeze.